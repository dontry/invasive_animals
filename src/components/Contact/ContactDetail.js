import React, { Fragment } from "react";
import styled from "styled-components";
//Material UI
import { green, grey, teal } from "material-ui/colors";
//Components
import { Paragraph, Title } from "../common/Text";

const ListWrapper = styled.ul`
  padding-top: 1rem;
  line-height: 1.5em;
  & li {
    margin-bottom: 2rem;
  }

  & * {
    color: ${green[50]} !important;
  }

  & li:hover {
    color: ${grey[50]} !important;
  }
`;

const ContactDetail = () => (
  <ListWrapper>
    <li>
      <Title variant="title" align="left" txtColor={grey[600]}>
        <i className="fas fa-phone" />
      </Title>
      <Paragraph variant="title" align="left" txtColor={grey[600]}>
        <a href="tel:+61 405 959 278">+61 405 959 278</a>
      </Paragraph>
    </li>
    <li>
      <Title variant="title" align="left" txtColor={grey[600]}>
        <i className="fas fa-envelope" />
      </Title>
      <Paragraph variant="title" align="left" txtColor={grey[600]}>
        <a href="mailto:vic.invasive@gmail.com">vic.invasive@gmail.com</a>
      </Paragraph>
    </li>
    <li>
      <Title variant="title" align="left" txtColor={grey[600]}>
        <i className="fab fa-facebook-square" />
      </Title>
      <Paragraph variant="title" align="left" txtColor={grey[600]}>
        <a href="https://wwww.facebook.com/Invasive-Defender-Vic-2160207437541280">
          Victorian Guardian
        </a>
      </Paragraph>
    </li>
    <li>
      <Title variant="title" align="left" txtColor={grey[600]}>
        <i class="fas fa-map-marker" />
      </Title>
      <Paragraph
        variant="title"
        align="left"
        txtColor={grey[600]}
        lineHeight="1.5em"
      >
        K, Monash University
        <br />
        Caulfield Campus, 900
        <br />
        Dandenong Road, Caulfield
        <br />
        East, Melbourne 3145
      </Paragraph>
    </li>
  </ListWrapper>
);

export default ContactDetail;
