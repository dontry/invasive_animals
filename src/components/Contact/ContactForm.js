import React, { Component, Fragment } from "react";
import PropTypes from "prop-types";
import styled from "styled-components";
import { red } from "material-ui/colors";
import { Field, reduxForm } from "redux-form";
import { Redirect, withRouter } from "react-router-dom";
import ReCAPTCHA from "react-google-recaptcha";
//Material UI
import Grid from "material-ui/Grid";
//Components
import { Mask } from "../common/Mask";
import Loader from "../common/Loader";
import { StyledButton } from "../common/ActionButtonGroup";
import { TextField } from "../common/FormFields";
import { Paragraph } from "../common/Text";
import MessageDialog from "../common/ConfirmationDialog";
import { validate } from "../../utils/formValidation";

const RECEPIENT = "vic.invasive@gmail.com";

export const StyledForm = styled.form`
  position: relative;
  padding: 2.5rem 2rem 3rem 0;
  width: 100%;
`;

const FormBody = styled(Grid)`
  && {
    padding-bottom: 2rem;
  }
`;

const FormFooter = styled(Grid)`
  && {
    width: 100%;
    text-align: right;
  }
`;

const Captcha = ({ input, meta: { touched, error } }) => (
  <Fragment>
    <Paragraph text_color={red[500]}>{touched && error}</Paragraph>
    <ReCAPTCHA
      sitekey="6LfMElYUAAAAAH8NaQnWfmsegkiaRjFqZ9Qq9ILi"
      onChange={res => input.onChange(res)}
    />
  </Fragment>
);

export class ContactForm extends Component {
  state = {
    dialogOpen: false,
    complete: false,
    message: ""
  };

  componentWillReceiveProps(nextProps) {
    if (
      nextProps.submitSucceeded &&
      (!nextProps.email.loading && !!nextProps.email.entity)
    ) {
      this.handleDialogOpen("Your report has been sent successfully");
    } else if (nextProps.submitSucceeded && nextProps.email.error) {
      this.handleDialogOpen(
        `The email delivery failed: ${JSON.stringify(nextProps.email.error)}`
      );
    }
  }

  componentWillUnmount() {
    this.props.reset();
  }

  handleSubmit = async email => {
    email = { ...email, to: RECEPIENT };
    await this.props.sendEmail(email);
  };

  handleDialogOpen = message => {
    this.setState({ dialogOpen: true, message });
  };

  handleDialogClose = () => {
    this.setState({ dialogOpen: false, complete: true });
  };

  render() {
    const { handleSubmit, submitting, submitSucceeded, email } = this.props;
    const { dialogOpen, message, complete } = this.state;

    if (complete === true) {
      return <Redirect to="/" />;
    }

    return (
      <Fragment>
        <StyledForm onSubmit={handleSubmit(this.handleSubmit)}>
          {(submitting || email.loading) && (
            <Mask>
              <Loader />
            </Mask>
          )}
          <FormBody container direction="row" justify="flex-start">
            <Field
              required
              component={TextField}
              name="username"
              label="Your name"
            />
            <Field
              required
              component={TextField}
              name="from"
              label="Your email address"
              type="email"
            />
            <Field
              required
              component={TextField}
              name="message"
              label="Message"
              multiline={true}
              witdh="100%"
              sm={12}
              fullWidth
              rows={5}
            />
          </FormBody>
          <FormFooter container direction="column" alignItems="flex-end">
            <Grid item>
              <Field name="recaptcha" component={Captcha} />
            </Grid>
            <br />
            <Grid item>
              <StyledButton type="primary">Send</StyledButton>
            </Grid>
          </FormFooter>
        </StyledForm>

        <MessageDialog
          open={dialogOpen}
          handleClose={this.handleDialogClose}
          message={message}
        />
      </Fragment>
    );
  }
}

export default reduxForm({ form: "contact", validate })(
  withRouter(ContactForm)
);
