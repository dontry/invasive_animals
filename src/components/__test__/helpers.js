import React from "react";
import defaultTheme from "../../assets/theme";
import { ThemeProvider } from "styled-components";

export function shallowWithTheme(component, theme = defaultTheme) {
  const context = shallow(<ThemeProvider theme={theme} />)
    .instance()
    .getChildContext();

  return shallow(component, { context });
}

export function mountWithTheme(component, theme = defaultTheme) {
  const context = shallow(<ThemeProvider theme={theme} />)
    .instance()
    .getChildContext();

  return mount(component, {
    context,
    childContextTypes: ThemeProvider.childContextTypes // Needed so child components receive the theme propd
  });
}
