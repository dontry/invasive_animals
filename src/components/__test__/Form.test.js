import React from "react";
import { ContactForm, StyledForm } from "../Contact/ContactForm";
import { shallowWithTheme } from "./helpers";
import { ExpansionPanelActions } from "material-ui";

describe("<ContactForm />", () => {
  let submitting, touched, error, reset, sendEmail, handleSubmit;
  const email = {
    username: "Trump",
    from: "dtrump2018@gmail.com",
    message: "MAGA",
    to: "vic.invasive@gmail.com"
  };
  beforeEach(() => {
    submitting = false;
    touched = false;
    error = null;
    reset = jest.fn();
    sendEmail = jest.fn();
    handleSubmit = fn => fn;
  });
  const buildForm = () => {
    const props = {
      submitting,
      fields: {
        username: {
          value: email.username,
          touched,
          error
        },
        from: {
          value: email.from,
          touched,
          error
        },
        message: {
          value: email.message,
          touched,
          error
        }
      },
      handleSubmit,
      reset,
      email: {
        entity: null,
        loading: false,
        error: null
      },
      sendEmail
    };
    return shallow(<ContactForm {...props} />);
  };

  // if use arrow function, then we have to use spyOn(instance(), fn)
  // to monitor function
  it("calls sendemail after submit; calls reset after unmount", () => {
    const wrapper = buildForm();
    wrapper.find(StyledForm).simulate("submit");

    expect(sendEmail.mock.calls.length).toBe(1);
    expect(toJson(wrapper)).toMatchSnapshot();
    wrapper.unmount();
    expect(reset.mock.calls.length).toBe(1);
  });

  it("calls handleSubmit", () => {
    const wrapper = buildForm();
    wrapper.instance().handleSubmit(email);

    expect(sendEmail).toHaveBeenCalledWith(email);
    wrapper.unmount();
  });
});
