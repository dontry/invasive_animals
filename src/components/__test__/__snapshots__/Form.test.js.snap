// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ContactForm /> calls sendemail after submit; calls reset after unmount 1`] = `
<React.Fragment>
  <styled.form
    onSubmit={[Function]}
  >
    <Styled(WithStyles(GridWrapper))
      container={true}
      direction="row"
      justify="flex-start"
    >
      <Field
        component={[Function]}
        label="Your name"
        name="username"
        required={true}
      />
      <Field
        component={[Function]}
        label="Your email address"
        name="from"
        required={true}
        type="email"
      />
      <Field
        component={[Function]}
        fullWidth={true}
        label="Message"
        multiline={true}
        name="message"
        required={true}
        rows={5}
        sm={12}
        witdh="100%"
      />
    </Styled(WithStyles(GridWrapper))>
    <Styled(WithStyles(GridWrapper))
      alignItems="flex-end"
      container={true}
      direction="column"
    >
      <WithStyles(GridWrapper)
        item={true}
      >
        <Field
          component={[Function]}
          name="recaptcha"
        />
      </WithStyles(GridWrapper)>
      <br />
      <WithStyles(GridWrapper)
        item={true}
      >
        <Styled(WithStyles(Button))
          type="primary"
        >
          Send
        </Styled(WithStyles(Button))>
      </WithStyles(GridWrapper)>
    </Styled(WithStyles(GridWrapper))>
  </styled.form>
  <MessageDialog
    handleClose={[Function]}
    message=""
    open={false}
  />
</React.Fragment>
`;
